src/Engine/EmitMSIL/CodeGenIL.cs

```diff
@@ -115,16 +113,182 @@ public CodeGenIL(IDictionary<string, IList> input, string filePath)
+            EmitOpCode(OpCodes.Newarr, typeof(T));
+
+            // Declare new array to store coerced values
+            var t = typeof(T).MakeArrayType();
+            var localBuilder = DeclareLocal(t);
+            var newArrIndex = localBuilder.LocalIndex;
+            EmitOpCode(OpCodes.Stloc, newArrIndex);
+
+            // Load array to be coerced.
+            EmitOpCode(OpCodes.Ldloc, localVarIndex);
+            // Emit new array
+            EmitOpCode(OpCodes.Ldloc, newArrIndex);
+
+            EmitOpCode(OpCodes.Ldc_I4_0);
+            mInfo = typeof(ICollection).GetMethod(nameof(ICollection.CopyTo));
+
+            // arr.CopyTo(newarr, 0);
+            EmitOpCode(OpCodes.Callvirt, mInfo);
+
+            EmitOpCode(OpCodes.Ldloc, newArrIndex);
+
+            return t;
+        }
+
+        private Type EmitArrayCoercion<S, T>()
+        {
+            // Find length for array to be coerced (already on top of eval stack), len
+            EmitOpCode(OpCodes.Ldlen);
+            EmitOpCode(OpCodes.Conv_I4);
+
+            // var newarr = new T[len];
+            EmitOpCode(OpCodes.Newarr, typeof(T));
+
+            // Declare new array to store coerced values
+            var t = typeof(T).MakeArrayType();
+            var localBuilder = DeclareLocal(t);
+            var newArrIndex = localBuilder.LocalIndex;
+            EmitOpCode(OpCodes.Stloc, newArrIndex);
+
+            // Emit for loop to loop over array and convert.
+
+            // i = 0;
+            var counterIndex = newArrIndex + 1;
+            EmitOpCode(OpCodes.Ldc_I4_0);
+            EmitOpCode(OpCodes.Stloc, counterIndex);
+
+            var loopBodyLabel = ilGen.DefineLabel();
+            var loopCondLabel = ilGen.DefineLabel();
+
+            EmitOpCode(OpCodes.Br_S, loopCondLabel);
+
+            // newarr[i] = (T)arr[i];
+            ilGen.MarkLabel(loopBodyLabel);
+
+            EmitOpCode(OpCodes.Ldloc, newArrIndex);
+            EmitOpCode(OpCodes.Ldloc, counterIndex);
+
+            // Load array to be coerced.
+            EmitOpCode(OpCodes.Ldloc, localVarIndex);
+            EmitOpCode(OpCodes.Ldloc, counterIndex);
+
+            if (typeof(S) == typeof(double))
+            {
+                EmitOpCode(OpCodes.Ldelem_R8);
+            }
+            else if (typeof(S) == typeof(long))
+            {
+                EmitOpCode(OpCodes.Ldelem_I8);
+            }
+            else if (typeof(S) == typeof(int))
+            {
+                EmitOpCode(OpCodes.Ldelem_I4);
+            }
+            if (typeof(T) == typeof(int))
+            {
+                EmitOpCode(OpCodes.Conv_I4);
+                EmitOpCode(OpCodes.Stelem_I4);
+            }
+            else if (typeof(T) == typeof(long))
+            {
+                EmitOpCode(OpCodes.Conv_I8);
+                EmitOpCode(OpCodes.Stelem_I8);
+            }
+            if (typeof(T) == typeof(double))
+            {
+                EmitOpCode(OpCodes.Conv_R8);
```

Comment:
is it safe to call this method if S is already a double? (I may not be understanding the intention here)

Reply:
Do you mean if this function is called as `EmitArrayCoercion<double, double>`?

Reply:
The intention here is basically to convert the source type (S) into the target (T) double type.